### PR TITLE
libfdisk based part plugin

### PR DESCRIPTION
**Pull request by vojtechtrefny**
Fixes: #245

So, this is finally it. After nearly 2 years we have the final version of the libfdisk rewrite of the part plugin.

I'll keep this as a draft for now. I plan to go through the code once more and make sure nothing got broken during the countless rebases etc. I also plan to go through the tests and make sure all corner cases are covered and also run the UDisks test suite with this version.

All parts of this PR have been reviewed before (except the last two commits I added today) but feel free to review this again. But this should really be the final version and I plan to fix smaller issues like #557 separately.

### CODE REVIEW

File: src/plugins/part.c [hunk 1]
```diff
@@ -467,65 +690,161 @@ static BDPartSpec* get_part_spec (PedDevice *dev, PedDisk *disk, PedPartition *p
...
         return NULL;
     }
 
-    ret = get_part_spec (dev, ped_disk, ped_part, error);
+    array = g_ptr_array_new_with_free_func ((GDestroyNotify) (void *) bd_part_spec_free);
 
-    /* the partition gets destroyed together with the disk */
-    ped_disk_destroy (ped_disk);
-    ped_device_destroy (dev);
+    while (fdisk_table_next_partition(table, itr, &pa) == 0) {
```
Comment (tbzatek) [inline src/plugins/part.c:797]: missing space :-p

File: src/plugins/part.c [hunk 2]
```diff
@@ -1000,107 +1085,370 @@ static PedPartition* add_part_to_disk (PedDevice *dev, PedDisk *disk, BDPartType
...
+        close_context (cxt);
         return NULL;
     }
 
-    succ = disk_commit (ped_disk, disk, error);
-    if (succ)
-        ret = get_part_spec (dev, ped_disk, ped_part, error);
+    if (fdisk_partition_has_partno (npa)) {
+        if (isdigit (disk[strlen(disk) - 1]))
+            ppath = g_strdup_printf ("%sp%zu", disk, fdisk_partition_get_partno (npa) + 1);
```
Comment (tbzatek) [inline src/plugins/part.c:1437]: Okay, I think this deserves more testing on _exotic_ drives: (e)MMC, NVMe, NVMeOF as the device node naming is rather non-traditional there.

File: src/plugins/part.c [hunk 3]
```diff
@@ -2007,6 +2902,9 @@ const gchar* bd_part_get_flag_str (BDPartFlag flag, GError **error) {
     return NULL;
 }
 
+/* string for BD_PART_TYPE_PROTECTED is "primary", that's what parted returns... */
+static const gchar*const part_types[6] = { "primary", "logical", "extended", "free", "metadata", "primary" };
```
Comment (tbzatek) [inline src/plugins/part.c:2906]: missing space.

File: src/plugins/part.c (hunk 2 above)
Comment (vojtechtrefny) [reply src/plugins/part.c:1437]: I think this should be OK for all devices. Yes, there are some weird names for NVMes and similar devices, but this should not affect partition names. The device name + number (or + "p" + number for device names ending with a digit) system should work everywhere -- it works this way even for "non-disk" devices like loop or md.

**Comment by vojtechtrefny** [comment]
Jenkins, test this please.

**Comment by vojtechtrefny** [comment]
Jenkins, test this please.

**Comment by vojtechtrefny** [comment]
I did one last fix and now all tests are passing except on CentOS 8 where we hit #557. I'm merging this now and I'll deal with the resize issue later.
